api/proquest/importer.py

```diff
@@ -789,61 +811,199 @@ def _parse_identifier(self, identifier):
+        """
+        for publication in self._get_publications(feed):
+            identifier = parse_identifier(self._db, publication.metadata.identifier)
+
+            feed_identifiers.append(identifier.identifier)
+
+    def _clean_removed_items(self, feed_identifiers):
+        """Make items that are no longer present in the ProQuest feed to be invisible in the CM's catalog.
+
+        :param feed_identifiers: List of identifiers present in the ProQuest feed
+        :type feed_identifiers: List[str]
+        """
+        self._logger.info(
+            "Started removing identifiers that are no longer present in the ProQuest feed"
+        )
+
+        items_to_remove = (
+            self._db.query(LicensePool)
+            .join(Collection)
+            .join(Identifier)
+            .filter(Collection.id == self.collection_id)
+            .filter(~Identifier.identifier.in_(feed_identifiers))
+        )
+
+        for item in items_to_remove:
+            item.unlimited_access = False
+
+        self._logger.info(
+            "Finished removing identifiers that are no longer present in the ProQuest feed"
+        )
+
+    def _get_feeds(self):
+        """Return a generator object traversing through a list of the ProQuest OPDS 2.0 feed pages.
+
+        :return: Generator object traversing through a list of the ProQuest OPDS 2.0 feed pages
+        :rtype: Iterable[opds2_ast.OPDS2Feed]
+        """
+        self._logger.info("Started fetching ProQuest paged OPDS 2.0 feeds")
+
+        feed_temporary_files = self._download_feed_pages()
+        page = 1
+        processed_number_of_items = 0
+        total_number_of_items = None
+
+        try:
+            self._logger.info("Started processing feed pages")
+
+            for feed_temporary_file in feed_temporary_files:
+                feed = self._parse_feed(page, feed_temporary_file)
+
+                # FIXME: We cannot short-circuit the feed import process
+                #  because ProQuest feed is not ordered by the publication's modified date.
+                #  This issue will be addressed in https://jira.nypl.org/browse/SIMPLY-3343
+                # if not self.feed_contains_new_data(feed):
+                #     break
+
+                if total_number_of_items is None:
+                    total_number_of_items = (
+                        feed.metadata.number_of_items
+                        if feed.metadata.number_of_items
+                        else 0
+                    )
 
-            # FIXME: We cannot short-circuit the feed import process
-            #  because ProQuest feed is not ordered by the publication's modified date.
-            #  This issue will be addressed in https://jira.nypl.org/browse/SIMPLY-3343
-            # if not self.feed_contains_new_data(feed):
-            #     break
+                processed_number_of_items += (
+                    feed.metadata.items_per_page if feed.metadata.items_per_page else 0
```

Comment:
This might indicate a problem with the OPDS2 parser rather than this code, but does this imply that every page of a paginated OPDS2 feed has the same number of items? That's not necessarily true.

Reply:
Or does `items_per_page` contain the number of items that were actually found on that particular page?

Reply:
Yes, it must contain the number of the items on a specific page.

Reply:
OK, I think the name could be improved but that's an issue for the parser library.